docs: policy control doc (#4104)
mesosphere/dcos-docs-site: pages/dkp/kommander/2.1/security/policy-controls/gatekeeper/index.md

---
layout: layout.pug
navigationTitle: Enforce policies using Gatekeeper
title: Enforce policies using Gatekeeper
menuWeight: 24
excerpt: Learn how to enforce policies using Gatekeeper
beta: false
enterprise: false
---

<!-- markdownlint-disable MD030 MD018-->

> [Gatekeeper][gatekeeper] is the policy controller for Kubernetes, allowing organizations to enforce configurable policies using the [Open Policy Agent][opa], a policy engine for Cloud Native environments hosted by CNCF as an incubation-level project.

Author: done

> This tutorial demonstrates how to use Gatekeeper to enforce policies by rejecting non-compliant resources. Specifically, this tutorial will demonstrate two constraints as a way to show how to use Gatekeeper as an alternative to [Pod Security Policies][psp]:

Reviewer: ```suggestion
This tutorial describes how to use Gatekeeper to enforce policies by rejecting non-compliant resources. Specifically, this tutorial describes two constraints as a way to show how to use Gatekeeper as an alternative to [Pod Security Policies][psp]:
```

Author: done

- [Prevent the running of privileged pods](#prevent-privileged-pods)
- [Prevent mounting host path volumes](#prevent-host-path-volumes)

## Before you begin

> Before starting this tutorial, you should verify the following:

Reviewer: ```suggestion
Before starting this tutorial, verify the following:
```

Author: done

- You must have access to a Linux, macOS, or Windows computer with a supported operating system version.

- You must have a properly deployed and running cluster. For information about deploying Kubernetes with default settings on different types of infrastructures, see the [Choose Infrastructure][choose-infrastructure] topic.

> - If you install Kommander with a custom configuration, make sure you enabled gatekeeper.

Author: @msdolbey @cmastr another pass please, the enable gatekeeper link I used is not accurate since it contains additional configs about http proxy..

Reviewer 2: Oh sure - no probs.

Reviewer 2: So, this reads fine enough for me, and I don't think there are any issues the way it's written, but more of a question to you @cmastr @cwyl02 - 
I think this is fine leaving it a bit up to the user to confirm, but have a quick question if we want to say something like:

You can confirm if the Gatekeeper app is running by running the command:

`kubectl get pods -A | grep gatekeeper`

Or something like that.

Reviewer 2: But, other than that, and I could go either way on that, this is good to me.

> - If you install Kommander with a custom configuration, make sure you enabled gatekeeper.

Reviewer 2: Caps for Gatekeeper.

```suggestion
- If you install Kommander with a custom configuration, make sure you enabled Gatekeeper.
```

> - Ensure Gatekeeper is [enabled](../../../install/http-proxy/#enable-gatekeeper)

Reviewer 3: since gatekeeper is enabled by default, I wonder if we can drop the link. Also, that page describes how to enable mutating webhooks which we don't really need for this doc (which describes validating webhooks i think)

Author: I think the link provides more info than we need. I dropped the link and changed the wording.

> - Ensure Gatekeeper is [enabled](../../../install/http-proxy/#enable-gatekeeper)

Reviewer: ```suggestion
- [Enable](../../../install/http-proxy/#enable-gatekeeper) Gatekeeper
```

Author: done

## Use Gatekeeper

Gatekeeper uses the [OPA Constraint Framework][opa-constraints] to describe and enforce policy. Before you can define a constraint, you must first define a `ConstraintTemplate`, which describes both the [`Rego`][opa-rego] (a powerful query language) that enforces the constraint and the schema of the constraint. The schema of the constraint allows an admin to fine-tune the behavior of a constraint, much like arguments to a function.

> The Gatekeeper repository includes a [library of policies][gatekeeper-psp] to replace Pod Security Policies which we will use in the following tutorials.

Reviewer 2: This link needs to be updated to ...  I'm pretty sure this one `[gatekeeper-library]` - which goes here: 
https://github.com/open-policy-agent/gatekeeper-library/tree/master/library/pod-security-policy

```suggestion
The Gatekeeper repository includes a [library of policies][gatekeeper-library] to replace Pod Security Policies which we will use in the following tutorials.
```

Author: done

### Prevent privileged pods

#### Define the ConstraintTemplate

Create the privileged pod policy constraint template `k8spspprivilegedcontainer` by running the following command:

```bash

> kubectl apply -f https://raw.githubusercontent.com/open-policy-agent/gatekeeper/master/library/pod-security-policy/privileged-containers/template.yaml

Reviewer 2: This got updated: 

```suggestion
kubectl apply -f https://raw.githubusercontent.com/open-policy-agent/gatekeeper-library/master/library/pod-security-policy/privileged-containers/template.yaml
```

Author: done

---
layout: layout.pug
navigationTitle: Enforce policies using Gatekeeper
title: Enforce policies using Gatekeeper
menuWeight: 24
excerpt: Learn how to enforce policies using Gatekeeper
beta: false
enterprise: false
---

<!-- markdownlint-disable MD024 -->

> [Gatekeeper][gatekeeper] is the policy controller for Kubernetes, allowing organizations to enforce configurable policies using the [Open Policy Agent][opa], a policy engine for Cloud Native environments hosted by CNCF as an incubation-level project.

Reviewer 2: ```suggestion
[Gatekeeper][gatekeeper] is the policy controller for Kubernetes, allowing organizations to enforce configurable policies using the [Open Policy Agent][opa], a policy engine for Cloud Native environments hosted by CNCF as a graduate-level project.
```

Author: done

This tutorial demonstrates how to use Gatekeeper to enforce policies by rejecting non-compliant resources. Specifically, this tutorial will demonstrate two constraints as a way to show how to use Gatekeeper as an alternative to [Pod Security Policies][psp]:

- [Prevent the running of privileged pods](#prevent-privileged-pods)
- [Prevent mounting host path volumes](#prevent-host-path-volumes)

## Before you begin

Before starting this tutorial, you should verify the following:

-   You must have access to a Linux, macOS, or Windows computer with a supported operating system version.

> -   You must have a properly deployed and running cluster. For information about deploying Kubernetes with default settings on different types of infrastructures, see the [Choose Infrastructure][choose-infrastructure].

Author: done
